### PR TITLE
Add `discussion-links-on-repo-lists` feature

### LINKED ISSUE #1130: Add links to issues and pull requests next to repo title
When working in organization with multiple repos... the org home page lists all the repos. clicking on the repo name will take to code and then we click on issues / pull requests. Being a maintainer, most of the times I work on issues / prs. It would be very handy if I can directly go on the repo's issues / prs page from the organization home page. 

One way to do is to have links to /issues and /pull requests next to the repo name itself.

Let me know your comments and suggestions. Would be happy to do a pr 😄

### PR DESCRIPTION
**Pull request by yakov116**
Closes #1130 

image

I followed https://github.com/sindresorhus/refined-github/issues/1130#issuecomment-529137735
> Honestly I think we can just replace that `X issues need help` and display it there instead. It's not a very useful copy now that everyone uses the label as _"up for grabs"_ rather than _"we actually need someone who knows more than us to fix it"_.
> 
> Plus I'd stick to the just show the icons, no counts. If anyone wants the counts, you can use **Mottie**'s userscript.

# Test

https://github.com/sindresorhus?tab=repositories&type=source
https://github.com/search?q=user%3Asindresorhus+something

<!--

Thanks for contributing! 🍄

1. LINKED ISSUES:
   Does this PR close/fix an existing issue? Write something like `Closes #10`

2. TEST URLS:
   Add some test URLs

3. SCREENSHOT:
   Add a screenshot here if your PR makes visual changes

### CODE REVIEW

**Comment by yakov116** [comment]
One issue. If the query changes then the feature is dropped

File: source/features/link-issues-pr-user-profile.tsx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+features.add({
+	id: __featureName__,
+	description: 'Adds a link to the issues and pull on the user profile repository tab.',
+	screenshot: 'https://user-images.githubusercontent.com/16872793/78670305-8ab3c780-78ab-11ea-9c17-4d07f124192e.png'
+}, {
+	include: [
+		features.isUserProfile
+	],
+	exclude: [
+		features.isOrganizationProfile
```
Comment (fregante) [inline source/features/link-issues-pr-user-profile.tsx:46]: Idea for a separate PR: Maybe isUserProfile should automatically exclude organizations. Other features should also be updated to reflect this change.

File: source/features/link-issues-pr-user-profile.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	],
+	load: features.nowAndOnAjaxedPages,
```
Comment (fregante) [inline source/features/link-issues-pr-user-profile.tsx:48]: This can't use `now` because it's applied on the whole page, not just one element. Skip the `elementReady` too

File: source/features/link-issues-pr-user-profile.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import React from 'dom-chef';
+import select from 'select-dom';
+import features from '../libs/features';
+import elementReady from 'element-ready';
+import issueIcon from 'octicon/issue-opened.svg';
+import pullRequestIcon from 'octicon/git-pull-request.svg';
+
+async function init(): Promise<false | void> {
+	const repositoryTab = await elementReady('[data-filterable-for="your-repos-filter"]');
+	if (!repositoryTab) {
```
Comment (fregante) [inline source/features/link-issues-pr-user-profile.tsx:10]: This specific condition could be part of `page-detect` because it can be detected by URL

Comment (yakov116) [reply source/features/link-issues-pr-user-profile.tsx:10]: So make a new one for it?

By the top repo we did this

**Comment by fregante** [comment]
> One issue. If the query changes then the feature is dropped

Which query?

> I dont know what to call the feature and can use some help. (Will add to readme once I have the name)

Maybe `discussion-links-on-repo-lists`, it can also apply to global searches https://github.com/search?q=xo+eslint

**Comment by yakov116** [comment]
> > One issue. If the query changes then the feature is dropped
> 
> Which query?
> 

Any change (The type, searching etc.)

**Comment by yakov116** [comment]
> Maybe `discussion-links-on-repo-lists`, it can also apply to global searches [github.com/search?q=xo+eslint](https://github.com/search?q=xo+eslint)

Total different layout, But I can add it

File: source/features/discussion-links-on-repo-lists.tsx [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+				{pullRequestIcon()}
+			</a>
+		);
+	}
+}
+
+function init(): void {
+	addIcons();
+	if (isUserProfileRepoTab()) {
+		observeElement('#user-repositories-list', addIcons);
```
Comment (fregante) [inline source/features/discussion-links-on-repo-lists.tsx:37]: Might as well use a second `feature.add` block for this ;)

Like

```js
, {
	include: [
		features.isUserProfileRepoTab,
		features.isGlobalSearchResults
	],
	load: features.onAjaxedPages,
	init
}, {
	include: [
		features.isUserProfileRepoTab
	],
	load: features.onAjaxedPages,
	init: () => observeElement('#user-repositories-list', addIcons);
}

Comment (yakov116) [reply source/features/discussion-links-on-repo-lists.tsx:37]: Cool!

File: source/features/discussion-links-on-repo-lists.tsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import pullRequestIcon from 'octicon/git-pull-request.svg';
+import observeElement from '../libs/simplified-element-observer';
+
+function addIcons(): void {
+	for (const repository of select.all('[itemprop="owns"]:not(.rgh-discussion-links), .repo-list-item')) {
+		repository.classList.add('rgh-discussion-links');
+		// Remove the help wanted issues
+		select('[href*="issues?q=label%3A%22help+wanted%"]', repository)?.remove();
+
+		const repositoryUrl = select<HTMLAnchorElement>('[itemprop="name codeRepository"], [data-hydro-click*="search_result.click"][href]', repository)!.href;
```
Comment (fregante) [inline source/features/discussion-links-on-repo-lists.tsx:14]: Ouch this selector. Any possible improvements?

File: source/features/discussion-links-on-repo-lists.tsx [hunk 6]
```diff
@@ -0,0 +1,52 @@
+import React from 'dom-chef';
+import select from 'select-dom';
+import features from '../libs/features';
+import issueIcon from 'octicon/issue-opened.svg';
+import {isUserProfileRepoTab} from '../libs/page-detect';
+import pullRequestIcon from 'octicon/git-pull-request.svg';
+import observeElement from '../libs/simplified-element-observer';
+
+function addIcons(): void {
+	for (const repository of select.all('[itemprop="owns"]:not(.rgh-discussion-links), .repo-list-item')) {
```
Comment (fregante) [inline source/features/discussion-links-on-repo-lists.tsx:9]: Can you select the repository link/name and then work back from there? A `repositoryLink.closest('x')` would probably end up being slightly safer/shorter

File: source/features/discussion-links-on-repo-lists.tsx (hunk 4 above)
Comment (yakov116) [reply source/features/discussion-links-on-repo-lists.tsx]: Typescript is not happy LOL

File: source/features/discussion-links-on-repo-lists.tsx [hunk 7]
```diff
@@ -0,0 +1,50 @@
+import React from 'dom-chef';
+import select from 'select-dom';
+import features from '../libs/features';
+import issueIcon from 'octicon/issue-opened.svg';
+import pullRequestIcon from 'octicon/git-pull-request.svg';
+import observeElement from '../libs/simplified-element-observer';
```
Comment (fregante) [inline source/features/discussion-links-on-repo-lists.tsx:6]: I'm not sure why this rule isn't active: https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/order.md

First sort by distance: node_modules, then same directory, then `..` last

And then in each of these 3 groups you can sort the "from" keyword

```suggestion
import React from 'dom-chef';
import select from 'select-dom';
import issueIcon from 'octicon/issue-opened.svg';
import pullRequestIcon from 'octicon/git-pull-request.svg';
import features from '../libs/features';
import observeElement from '../libs/simplified-element-observer';
```

File: source/features/discussion-links-on-repo-lists.tsx [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+		features.isGlobalSearchResults
+	],
+	load: features.onAjaxedPages,
+	init
+}, {
+	include: [
+		features.isUserProfileRepoTab
+	],
+	load: features.onAjaxedPages,
+	init: () => observeElement('#user-repositories-list', init)
```
Comment (fregante) [inline source/features/discussion-links-on-repo-lists.tsx:49]:
```suggestion
	init: () => observeElement('#user-repositories-list', init) ?? false
```

File: source/features/discussion-links-on-repo-lists.tsx (hunk 6 above)
Comment (fregante) [reply source/features/discussion-links-on-repo-lists.tsx:9]:
```suggestion
	const repositories = select.all([
			'[itemprop="name codeRepository"]', // `isUserProfileRepoTab`
			'[data-hydro-click*=\'"model_name":"Repository"\']' // `isGlobalSearchResults`
	]);
	for (const repository of repositories) {
```

File: source/features/discussion-links-on-repo-lists.tsx (hunk 5 above)
Comment (fregante) [reply source/features/discussion-links-on-repo-lists.tsx:14]: Yes: https://github.com/sindresorhus/refined-github/pull/2981/files#r405112846

**Comment by yakov116** [comment]
>Can you select the repository link/name and then work back from there? A repositoryLink.closest('li') would probably end up being slightly safer/shorter

Where for the global search?

The way I did it it can work for both the Search and profiles

File: source/features/discussion-links-on-repo-lists.tsx [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+		const repository = repositoryLink.closest('li')!;
+		// Remove the help wanted issues
+		select('[href*="issues?q=label%3A%22help+wanted%"]', repository)?.remove();
+
+		const repositoryUrl = select<HTMLAnchorElement>('[itemprop="name codeRepository"], [data-hydro-click*="search_result.click"][href]', repository)!.href;
+		// Need to use previousSibling since the updated text is not in an element
+		select('relative-time', repository)!.previousSibling!.before(
+			<a
+				className="muted-link mr-3"
+				href={repositoryUrl + '/issues?q=is%3Aissue+is%3Aopen'}
```
Comment (fregante) [inline source/features/discussion-links-on-repo-lists.tsx:26]:
```suggestion
				href={repositoryLink.href + '/issues?q=is%3Aissue+is%3Aopen'}
```

File: source/features/discussion-links-on-repo-lists.tsx [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+			>
+				{issueIcon()}
+			</a>,
+			<a
+				className="muted-link mr-3"
+				href={repositoryUrl + '/pulls?q=is%3Aissue+is%3Aopen'}
```
Comment (fregante) [inline source/features/discussion-links-on-repo-lists.tsx:32]:
```suggestion
				href={repositoryLink.href + '/pulls?q=is%3Aissue+is%3Aopen'}
```

File: source/features/discussion-links-on-repo-lists.tsx [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+	]);
+
+	for (const repositoryLink of repositories) {
+		repositoryLink.classList.add('rgh-discussion-links');
+
+		const repository = repositoryLink.closest('li')!;
+		// Remove the help wanted issues
+		select('[href*="issues?q=label%3A%22help+wanted%"]', repository)?.remove();
+
+		const repositoryUrl = select<HTMLAnchorElement>('[itemprop="name codeRepository"], [data-hydro-click*="search_result.click"][href]', repository)!.href;
```
Comment (fregante) [inline source/features/discussion-links-on-repo-lists.tsx:21]:
```suggestion
```

File: source/libs/page-detect.ts [hunk 12]
```diff
@@ -373,6 +373,11 @@ export const _isBranches = [
 export const isUserProfile = (): boolean => select.exists('.user-profile-nav');
 export const _isUserProfile = domBased;
 
+export const isUserProfileRepoTab = (): boolean => location.search.startsWith('?tab=repositories');
```
Comment (fregante) [inline source/libs/page-detect.ts:376]: This isn't enough. It even matches `https://github.com/?tab=repositories`

Comment (yakov116) [reply source/libs/page-detect.ts:377]: When I wrote originally I added isuserprofile but I did not know how to test it

File: source/libs/page-detect.ts [hunk 13]
```diff
@@ -374,7 +374,7 @@ export const _isBranches = [
 export const isUserProfile = (): boolean => select.exists('.user-profile-nav');
 export const _isUserProfile = domBased;
 
-export const isUserProfileRepoTab = (): boolean => location.search.startsWith('?tab=repositories');
+export const isUserProfileRepoTab = (): boolean => location.search.startsWith('?tab=repositories') && location.pathname.length > 1;
```
Comment (fregante) [inline source/libs/page-detect.ts:377]: ?

This needs to include isUserProfile.

Comment (yakov116) [reply source/libs/page-detect.ts:377]: Working on it

File: source/features/discussion-links-on-repo-lists.tsx [hunk 14]
```diff
@@ -20,7 +20,7 @@ function init(): void {
 		select('[href*="issues?q=label%3A%22help+wanted"]', repository)?.remove();
 
 		// Place before the "Updated on" element
-		select('relative-time', repository)!.parentElement!.before(
+		select('relative-time', repository)!.previousSibling!.before(
```
Comment (yakov116) [inline source/features/discussion-links-on-repo-lists.tsx:23]: parent element on work on the search tab not the user profile tab.

image

Comment (fregante) [reply source/features/discussion-links-on-repo-lists.tsx:23]: It looks like the DOM is different:

- global search wraps the time: https://github.com/search?q=user%3Asindresorhus+is

global

- user profile doesn't: https://github.com/sindresorhus?tab=repositories&type=source

user profile

`previousSibling` produces the wrong DOM in the global search, but maybe it's fine. The alternative may be too wordy.

File: source/libs/page-detect.ts [hunk 15]
```diff
@@ -373,6 +374,14 @@ export const _isBranches = [
 export const isUserProfile = (): boolean => select.exists('.user-profile-nav');
 export const _isUserProfile = domBased;
 
+export const isUserProfileRepoTab = (): boolean => isUserProfile() &&
+	new URLSearchParams(location.search).get('tab') === 'repositories' &&
+	location.pathname !== '/'; // Gotcha for 'https://github.com/?tab=repositories'
```
Comment (fregante) [inline source/libs/page-detect.ts:379]: ?

This is already covered by isUserProfile

Comment (yakov116) [reply source/libs/page-detect.ts:379]: I know. My AVA is not working locally (And my computer is doing funny things) I Forgot to remove it

Comment (fregante) [reply source/libs/page-detect.ts]: What's not working? Make sure you test this on Node 12.15. You can use [n](https://www.npmjs.com/package/n) to switch between versions

Comment (yakov116) [reply source/libs/page-detect.ts]: 'TS_NODE_FILES' is not recognized as an internal or external command

Comment (fregante) [reply source/libs/page-detect.ts]: Are you on Windows 10? It should work on bash

Comment (yakov116) [reply source/libs/page-detect.ts]: > Are you on Windows 10? It should work on bash

Yes. Let me try

Comment (fregante) [reply source/libs/page-detect.ts]: Or, more easily for now, first set the `TS_NODE_FILES` environment variable and then just run `npx ava` directly, not `npm run test:js`

Comment (yakov116) [reply source/libs/page-detect.ts]:
```console
ava

  
  × Timed out while running tests

  × No tests found in test\search-query.ts
  × No tests found in test\page-detect.ts
  × No tests found in test\utils.ts

```

Give on trying to fix it

**Comment by yakov116** [comment]
What is wrong? Why does the test not pass?

File: source/libs/page-detect.ts [hunk 16]
```diff
@@ -373,6 +374,13 @@ export const _isBranches = [
 export const isUserProfile = (): boolean => select.exists('.user-profile-nav');
 export const _isUserProfile = domBased;
 
+export const isUserProfileRepoTab = (): boolean => isUserProfile() &&
+	new URLSearchParams(location.search).get('tab') === 'repositories';
+export const _isUserProfileRepoTab = [
+	'https://github.com/sindresorhus?tab=repositories&type=source',
+	'https://github.com/sindresorhus?type=source&tab=repositories'
+];
```
Comment (fregante) [inline source/libs/page-detect.ts:382]:
```suggestion
export const _isUserProfileRepoTab = domBased;
```

File: source/libs/page-detect.ts [hunk 17]
```diff
@@ -373,6 +374,13 @@ export const _isBranches = [
 export const isUserProfile = (): boolean => select.exists('.user-profile-nav');
 export const _isUserProfile = domBased;
 
+export const isUserProfileRepoTab = (): boolean => isUserProfile() &&
+	new URLSearchParams(location.search).get('tab') === 'repositories';
```
Comment (fregante) [inline source/libs/page-detect.ts:378]:
```suggestion
export const isUserProfileRepoTab = (): boolean =>
	isUserProfile() &&
	new URLSearchParams(location.search).get('tab') === 'repositories';
```

**Comment by yakov116** [comment]
> I was wondering myself.
> 
> And, yeah, `isUserProfile` is DOM-based so that makes this test DOM-based as well.

I should have explained better here. https://github.com/sindresorhus/refined-github/pull/2981#discussion_r405608813

**Comment by yakov116** [comment]
> I was wondering myself.
> 
> And, yeah, `isUserProfile` is DOM-based so that makes this test DOM-based as well.

I should have explained better here. https://github.com/sindresorhus/refined-github/pull/2981#discussion_r405608813

File: source/libs/page-detect.ts (hunk 15 above)
Comment (fregante) [reply source/libs/page-detect.ts]: This happens to me sometimes too, it's because ava tries to run all the tests in parallel and times out. Sometimes I just try again and it works.

Try `ava --serial`.
